File: lib/api/community.ts
// API utilities for community forum
import { fetchWithAuth } from '@/lib/api-utils';

export interface CommunityPost {
  id: string;
  title: string;
  content: string;
  type: 'question' | 'discussion' | 'study-group' | 'resource-sharing' | 'help-wanted' | 'deleted';
  author: {
    id: string;
    name: string;
    role: 'student' | 'ta' | 'tutor' | 'mentor';
    avatar?: string;
    reputation?: number;
  };
  tags: string[];
  isPinned: boolean;
  isResolved: boolean;
  hasAcceptedAnswer: boolean;
  upvotes: number;
  downvotes: number;
  answerCount: number;
  viewCount: number;
  createdAt: string;
  updatedAt: string;
  lastActivity: string;
}

export interface CommunityAnswer {
  id: string;
  postId: string;
  content: string;
  latexBlocks?: string[];
  author: {
    id: string;
    name: string;
    role: 'student' | 'ta' | 'tutor' | 'mentor';
    avatar?: string;
    reputation?: number;
  };
  isAccepted: boolean;
  isHelpful: boolean;
  upvotes: number;
  downvotes: number;
  createdAt: string;
  updatedAt: string;
}

export interface CreatePostData {
  title: string;
  content: string;
  type: CommunityPost['type'];
  tags: string[];
  isPinned?: boolean;
}

export interface CreateAnswerData {
  content: string;
  latexBlocks?: string[];
}

export interface CommunityPostsResponse {
  posts: CommunityPost[];
  pagination: {
    page: number;
    per_page: number;
    total: number;
    pages: number;
  };
}

export interface CommunityPostDetailResponse {
  post: CommunityPost;
  answers: CommunityAnswer[];
}

export const communityApi = {
  // Get all posts for a course
  async getPosts(
    courseId: string, 
    options: {
      page?: number;
      per_page?: number;
      sort?: 'recent' | 'popular' | 'answered';
      type?: CommunityPost['type'];
      include_deleted?: boolean;
      search?: string;
    } = {}
  ): Promise<CommunityPostsResponse> {
    const params = new URLSearchParams();
    if (options.page) params.append('page', options.page.toString());
    if (options.per_page) params.append('per_page', options.per_page.toString());
    if (options.sort) params.append('sort', options.sort);
    if (options.type) params.append('type', options.type);
    if (options.include_deleted) params.append('include_deleted', 'true');
    if (options.search) params.append('search', options.search);

    const queryString = params.toString();
    const url = `/api/courses/${courseId}/community/posts${queryString ? `?${queryString}` : ''}`;
    
    return fetchWithAuth(url, {
      method: 'GET',
    });
  },

  // Create a new post
  async createPost(courseId: string, postData: CreatePostData): Promise<{ message: string; post: CommunityPost }> {
    return fetchWithAuth(`/api/courses/${courseId}/community/posts`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(postData),
    });
  },

  // Get a specific post with answers
  async getPost(courseId: string, postId: string): Promise<CommunityPostDetailResponse> {
    return fetchWithAuth(`/api/courses/${courseId}/community/posts/${postId}`, {
      method: 'GET',
    });
  },

  // Create an answer for a post
  async createAnswer(
    courseId: string, 
    postId: string, 
    answerData: CreateAnswerData
  ): Promise<{ message: string; answer: CommunityAnswer }> {
    return fetchWithAuth(`/api/courses/${courseId}/community/posts/${postId}/answers`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(answerData),
    });
  },

  // Vote on a post
  async votePost(
    courseId: string, 
    postId: string, 
    voteType: 'upvote' | 'downvote'
  ): Promise<{ message: string; upvotes: number; downvotes: number }> {
    return fetchWithAuth(`/api/courses/${courseId}/community/posts/${postId}/vote`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ voteType }),
    });
  },

  // Vote on an answer
  async voteAnswer(
    courseId: string, 
    answerId: string, 
    voteType: 'upvote' | 'downvote'
  ): Promise<{ message: string; upvotes: number; downvotes: number }> {
    return fetchWithAuth(`/api/courses/${courseId}/community/answers/${answerId}/vote`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ voteType }),
    });
  },

  // Track a view for a post
  async trackView(
    courseId: string, 
    postId: string, 
    sessionId: string
  ): Promise<{ message: string; viewCount: number }> {
    return fetchWithAuth(`/api/courses/${courseId}/community/posts/${postId}/view`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ sessionId }),
    });
  },

  // Get user's vote for a post
  async getUserPostVote(
    courseId: string, 
    postId: string
  ): Promise<{ vote: 'upvote' | 'downvote' | null }> {
    return fetchWithAuth(`/api/courses/${courseId}/community/posts/${postId}/user-vote`, {
      method: 'GET',
    });
  },

  // Get user's vote for an answer
  async getUserAnswerVote(
    courseId: string, 
    answerId: string
  ): Promise<{ vote: 'upvote' | 'downvote' | null }> {
    return fetchWithAuth(`/api/courses/${courseId}/community/answers/${answerId}/user-vote`, {
      method: 'GET',
    });
  },

  // Edit a post
  async editPost(
    courseId: string, 
    postId: string, 
    postData: Partial<CreatePostData>
  ): Promise<{ message: string; post: CommunityPost }> {
    return fetchWithAuth(`/api/courses/${courseId}/community/posts/${postId}`, {
      method: 'PUT',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(postData),
    });
  },

  // Delete a post
  async deletePost(
    courseId: string, 
    postId: string
  ): Promise<{ message: string; deleted: boolean; soft_delete: boolean }> {
    return fetchWithAuth(`/api/courses/${courseId}/community/posts/${postId}`, {
      method: 'DELETE',
    });
  },

  // Delete an answer
  async deleteAnswer(
    courseId: string, 
    answerId: string
  ): Promise<{ message: string; deleted: boolean }> {
    return fetchWithAuth(`/api/courses/${courseId}/community/answers/${answerId}`, {
      method: 'DELETE',
    });
  },
};

File: backend/app/models/community.py
import uuid
from datetime import datetime
from app.init import db
from sqlalchemy.orm import relationship
from sqlalchemy import Text, JSON


class CommunityPost(db.Model):
    __tablename__ = 'users_courses_community'

    id = db.Column(db.String, primary_key=True, default=lambda: str(uuid.uuid4()))
    course_id = db.Column(db.String, db.ForeignKey('courses.combo_id'), nullable=False, index=True)
    user_id = db.Column(db.String, db.ForeignKey('users.id'), nullable=False, index=True)
    
    # Post content
    title = db.Column(db.String(200), nullable=False)
    content = db.Column(Text, nullable=False)  # Rich HTML content from TiptapEditor
    type = db.Column(db.Enum('question', 'discussion', 'study-group', 'resource-sharing', 'help-wanted', 'deleted', name='post_type_enum'), nullable=False)
    
    # Post metadata
    tags = db.Column(JSON, default=list)  # Store as JSON array
    is_pinned = db.Column(db.Boolean, default=False)
    is_resolved = db.Column(db.Boolean, default=False)
    has_accepted_answer = db.Column(db.Boolean, default=False)
    
    # Engagement metrics
    upvotes = db.Column(db.Integer, default=0)
    downvotes = db.Column(db.Integer, default=0)
    view_count = db.Column(db.Integer, default=0)
    
    # Timestamps
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    last_activity = db.Column(db.DateTime, default=datetime.utcnow)
    
    # Relationships
    user = relationship("User", backref="community_posts")
    course = relationship("Course", backref="community_posts")
    answers = relationship("CommunityAnswer", back_populates="post", cascade="all, delete-orphan")
    votes = relationship("CommunityPostVote", back_populates="post", cascade="all, delete-orphan")
    post_views = relationship("CommunityPostView", backref="post", cascade="all, delete-orphan")

    def to_dict(self):
        return {
            'id': self.id,
            'title': self.title,
            'content': self.content,
            'type': self.type,
            'author': {
                'id': self.user.id,
                'name': self.user.name,
                'role': 'student',  # Default role, can be enhanced later
            },
            'tags': self.tags or [],
            'isPinned': self.is_pinned,
            'isResolved': self.is_resolved,
            'hasAcceptedAnswer': self.has_accepted_answer,
            'upvotes': self.upvotes,
            'downvotes': self.downvotes,
            'answerCount': len(self.answers),
            'viewCount': self.view_count,
            'createdAt': self.created_at.isoformat() + 'Z',
            'updatedAt': self.updated_at.isoformat() + 'Z',
            'lastActivity': self.last_activity.isoformat() + 'Z',
        }


class CommunityAnswer(db.Model):
    __tablename__ = 'users_courses_community_answers'

    id = db.Column(db.String, primary_key=True, default=lambda: str(uuid.uuid4()))
    post_id = db.Column(db.String, db.ForeignKey('users_courses_community.id'), nullable=False, index=True)
    user_id = db.Column(db.String, db.ForeignKey('users.id'), nullable=False, index=True)
    
    # Answer content
    content = db.Column(Text, nullable=False)  # Rich HTML content from TiptapEditor
    latex_blocks = db.Column(JSON, default=list)  # Store LaTeX blocks separately for processing
    
    # Answer metadata
    is_accepted = db.Column(db.Boolean, default=False)
    is_helpful = db.Column(db.Boolean, default=False)
    
    # Engagement metrics
    upvotes = db.Column(db.Integer, default=0)
    downvotes = db.Column(db.Integer, default=0)
    
    # Timestamps
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    updated_at = db.Column(db.DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)
    
    # Relationships
    post = relationship("CommunityPost", back_populates="answers")
    user = relationship("User", backref="community_answers")
    votes = relationship("CommunityAnswerVote", back_populates="answer", cascade="all, delete-orphan")

    def to_dict(self):
        return {
            'id': self.id,
            'postId': self.post_id,
            'content': self.content,
            'latexBlocks': self.latex_blocks or [],
            'author': {
                'id': self.user.id,
                'name': self.user.name,
                'role': 'student',  # Default role
                'reputation': 0  # Can be calculated later
            },
            'isAccepted': self.is_accepted,
            'isHelpful': self.is_helpful,
            'upvotes': self.upvotes,
            'downvotes': self.downvotes,
            'createdAt': self.created_at.isoformat() + 'Z',
            'updatedAt': self.updated_at.isoformat() + 'Z',
        }


class CommunityPostVote(db.Model):
    __tablename__ = 'users_courses_community_post_votes'

    id = db.Column(db.String, primary_key=True, default=lambda: str(uuid.uuid4()))
    post_id = db.Column(db.String, db.ForeignKey('users_courses_community.id'), nullable=False)
    user_id = db.Column(db.String, db.ForeignKey('users.id'), nullable=False)
    vote_type = db.Column(db.Enum('upvote', 'downvote', name='vote_type_enum'), nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    
    # Relationships
    post = relationship("CommunityPost", back_populates="votes")
    user = relationship("User", backref="community_post_votes")
    
    # Ensure one vote per user per post
    __table_args__ = (db.UniqueConstraint('post_id', 'user_id', name='unique_post_vote'),)


class CommunityAnswerVote(db.Model):
    __tablename__ = 'users_courses_community_answer_votes'

    id = db.Column(db.String, primary_key=True, default=lambda: str(uuid.uuid4()))
    answer_id = db.Column(db.String, db.ForeignKey('users_courses_community_answers.id'), nullable=False)
    user_id = db.Column(db.String, db.ForeignKey('users.id'), nullable=False)
    vote_type = db.Column(db.Enum('upvote', 'downvote', name='answer_vote_type_enum'), nullable=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    
    # Relationships
    answer = relationship("CommunityAnswer", back_populates="votes")
    user = relationship("User", backref="community_answer_votes")
    
    # Ensure one vote per user per answer
    __table_args__ = (db.UniqueConstraint('answer_id', 'user_id', name='unique_answer_vote'),)


class CommunityPostView(db.Model):
    __tablename__ = 'users_courses_community_post_views'

    id = db.Column(db.String, primary_key=True, default=lambda: str(uuid.uuid4()))
    post_id = db.Column(db.String, db.ForeignKey('users_courses_community.id'), nullable=False)
    user_id = db.Column(db.String, db.ForeignKey('users.id'), nullable=False)
    session_id = db.Column(db.String, nullable=False)  # Track by session to handle page reloads
    created_at = db.Column(db.DateTime, default=datetime.utcnow)
    
    # Relationships
    user = relationship("User", backref="community_post_views")
    
    # Ensure one view per user per session per post
    __table_args__ = (db.UniqueConstraint('post_id', 'user_id', 'session_id', name='unique_post_view'),)
